Date: Fri, 12 Apr 2002 13:44:22 +0200
From: Marco d'Itri <md@Linux.IT>
To: mbp@samba.org
Subject: rsync and debian mirrors

I run one of the debian mirrors, and I had to write this patch because
my archive is split between more than one disk. Would you accept a more
polished version of this patch for inclusion in rsync?

[Updated to latest CVS source by Wayne Davison.]

--- syscall.c	18 Feb 2004 22:33:21 -0000	1.30
+++ syscall.c	22 Apr 2004 23:48:45 -0000
@@ -58,9 +58,14 @@ int do_symlink(char *fname1, char *fname
 #if HAVE_LINK
 int do_link(char *fname1, char *fname2)
 {
+	int st;
+
 	if (dry_run) return 0;
 	RETURN_ERROR_IF_RO_OR_LO;
-	return link(fname1, fname2);
+	st = link(fname1, fname2);
+	if (/*soften_links &&*/ st != 0 && errno == EXDEV)
+		st = symlink(fname1, fname2);
+	return st;
 }
 #endif
 
